File: ForgeEngine/Sources/ForgeEngine/DeckValidator.swift
import Foundation

/// Las reglas duras que Forge NUNCA puede violar.
/// Espejo 1:1 de `engine-reference/forge/validator.py`.
///
/// Regla de producto: si la colección no permite cumplirlas, Forge avisa y
/// explica el compromiso en lugar de generar un mazo defectuoso.
public enum DeckValidator {
    public static let basicLands: Set<String> = ["Plains", "Island", "Swamp", "Mountain", "Forest"]
    public static let maxCopies = 4

    public struct Violation: Equatable, CustomStringConvertible {
        public enum Kind { case notOwned, tooManyCopies, wrongSize, colorIdentity, landCount, curveOutOfRange }
        public let kind: Kind
        public let message: String
        public var description: String { message }
    }

    public static func validate(deck: Deck, pool: [String: Card]) -> [Violation] {
        var violations: [Violation] = []
        let everything = deck.cards.merging(deck.lands, uniquingKeysWith: +)

        // 1-2: propiedad y límite de copias
        for (name, qty) in everything {
            guard let owned = pool[name] else {
                violations.append(.init(kind: .notOwned, message: "'\(name)' no está en la colección"))
                continue
            }
            if qty > owned.qty {
                violations.append(.init(kind: .notOwned, message: "'\(name)': usa \(qty), posee \(owned.qty)"))
            }
            if qty > maxCopies && !basicLands.contains(name) {
                violations.append(.init(kind: .tooManyCopies, message: "'\(name)': \(qty) copias supera el límite de \(maxCopies)"))
            }
        }

        // 3: tamaño exacto
        if deck.totalCards != ManaCurve.deckSize {
            violations.append(.init(kind: .wrongSize, message: "el mazo tiene \(deck.totalCards) cartas, deben ser \(ManaCurve.deckSize)"))
        }

        // 4: identidad de color
        let allowed = Set(deck.colors)
        for name in deck.cards.keys {
            guard let card = pool[name] else { continue }
            if !Set(card.colors).isSubset(of: allowed) {
                violations.append(.init(kind: .colorIdentity, message: "'\(name)' (\(card.colors)) fuera de la identidad \(deck.colors)"))
            }
        }

        // 5: tierras y coste medio dentro del arquetipo
        let landCount = deck.lands.values.reduce(0, +)
        if !deck.archetype.landRange.contains(landCount) {
            violations.append(.init(kind: .landCount, message: "\(landCount) tierras fuera del rango \(deck.archetype.landRange) de \(deck.archetype.rawValue)"))
        }
        let knownCards = deck.cards.filter { pool[$0.key] != nil }
        let avg = ManaCurve.averageCMC(cards: knownCards, pool: pool)
        if !knownCards.isEmpty && !deck.archetype.averageCMCRange.contains(avg) {
            violations.append(.init(kind: .curveOutOfRange, message: String(format: "coste medio %.2f fuera del rango de %@", avg, deck.archetype.rawValue)))
        }

        return violations
    }
}

File: forge_engine/lib/src/deck_validator.dart
import 'mana_curve.dart';
import 'models.dart';

/// Las reglas duras que Forge NUNCA puede violar.
/// Espejo 1:1 de `engine-reference/forge/validator.py`.
///
/// Regla de producto: si la colección no permite cumplirlas, Forge avisa y
/// explica el compromiso en lugar de generar un mazo defectuoso.
class DeckValidator {
  static const basicLands = {'Plains', 'Island', 'Swamp', 'Mountain', 'Forest'};
  static const maxCopies = 4;

  /// Devuelve la lista de violaciones (vacía = mazo válido).
  static List<String> validate(Deck deck, Map<String, Card> pool) {
    final errors = <String>[];
    final everything = <String, int>{...deck.cards};
    deck.lands.forEach((k, v) => everything[k] = (everything[k] ?? 0) + v);

    // 1-2: propiedad y límite de copias
    everything.forEach((name, qty) {
      final owned = pool[name];
      if (owned == null) {
        errors.add("'$name' no está en la colección");
        return;
      }
      if (qty > owned.qty) {
        errors.add("'$name': usa $qty, posee ${owned.qty}");
      }
      if (qty > maxCopies && !basicLands.contains(name)) {
        errors.add("'$name': $qty copias supera el límite de $maxCopies");
      }
    });

    // 3: tamaño exacto
    if (deck.totalCards != ManaCurve.deckSize) {
      errors.add(
          'el mazo tiene ${deck.totalCards} cartas, deben ser ${ManaCurve.deckSize}');
    }

    // 4: identidad de color
    final allowed = deck.colors.split('').toSet();
    for (final name in deck.cards.keys) {
      final card = pool[name];
      if (card == null) continue;
      if (card.colors.split('').toSet().difference(allowed).isNotEmpty) {
        errors.add(
            "'$name' (${card.colors}) fuera de la identidad ${deck.colors}");
      }
    }

    // 5: tierras y coste medio dentro del arquetipo
    final landCount = deck.lands.values.fold(0, (a, b) => a + b);
    if (landCount < deck.archetype.landMin ||
        landCount > deck.archetype.landMax) {
      errors.add(
          '$landCount tierras fuera del rango ${deck.archetype.landMin}-${deck.archetype.landMax} de ${deck.archetype.name}');
    }
    final knownCards = Map<String, int>.fromEntries(
        deck.cards.entries.where((e) => pool.containsKey(e.key)));
    if (knownCards.isNotEmpty) {
      final avg = ManaCurve.averageCmc(knownCards, pool);
      if (avg < deck.archetype.cmcMin || avg > deck.archetype.cmcMax) {
        errors.add(
            'coste medio ${avg.toStringAsFixed(2)} fuera del rango de ${deck.archetype.name}');
      }
    }

    return errors;
  }
}
